tools/keepalived/keepalived/check/ipwrapper.c

```diff
@@ -1024,6 +1024,12 @@ clear_diff_rs(virtual_server_t *old_vs, virtual_server_t *new_vs, list old_check
 		}
 	}
 	clear_service_rs(old_vs, rs_to_remove, false);
+
+	//keep new_vs quorum_state_up same with old_vs
+	if (old_vs->quorum_state_up != new_vs->quorum_state_up) {
+		new_vs->quorum_state_up = old_vs->quorum_state_up;
```

Comment:
After clear_diff_services (ipwrapper.c:1377) in clear_diff_rs, old_vs->quorum_state_up maybe changed.
For example, old vip 10.1.2.100 -> old rs 192.168.1.10, after reload vip 10.1.2.100 -> new rs 192.168.2.10. in clear_diff_rs, 192.168.1.10 is the last rs in old_vs, after remove it ,the old_vs will be down in clear_service_rs(ipwrapper.c:292); but new_vs still up, because new rs 192.168.2.10 is up.

Reply:
Yeah, you are totally right. But according to my tests of this patch, quorum_down and quorum_up scripts are executed almost simutaneously, and sometimes  in reverse sequence. Could you reproduce this problem.

Reply:
quorum scripts executed sequence decided by old_vs and new_vs state.
For example, 
1) old_vs before reload 10.1.2.100 -> 192.168.1.10(rs is up,so old vip quorum_state_up is 1), new_vs after reload 10.1.2.100 -> 192.168.2.10(rs is up,so new vip quorum_state_up is 1), in this patch the new_vs state will change from down(ipwrapper.c:292) to up in perform_svr_state(ipwrapper.c:619), quorum_down scripts will be excuted first and then quorum_up scripts in a short period of time.
2) old_vs before reload 10.1.2.100 -> 192.168.1.10(rs is down, so old vip quorum_state_up is 0),new_vs after reload 10.1.2.100 -> 192.168.2.10 (rs is down, but in reload rs state is up in default, after 3 times healthcheck rs state change to down); in this patch the new_vs state is the same with old_vs(down) in clear_diff_services, it changes to up(excuted quorum_up scripts) in perform_svr_state (ipwrapper.c:619) because rs is default up, after healthcheck rs is down,  it changes to down(excuted quorum_down scripts).
Are your tests include the 2 conditions？

Reply:
Only the first case tested, where the final state should be UP but sometimes reaches DOWN. You may reproduce it with config similar as below.

```
virtual_server  group test {
    delay_loop 3
    lb_algo rr
    lb_kind FNAT
    protocol TCP
    
    laddr_group_name laddr_g1
    quorum_up "/tmp/ts.sh UP"
    quorum_down "/tmp/ts.sh DOWN"
    
#    real_server <RS_1> 80 {
#        weight 100
#        
#        inhibit_on_failure
#        TCP_CHECK {
#            retry 1
#            connect_timeout 1
#            connect_port 80
#        }
#    }

   real_server <RS_2> 80 {
        weight 100
        
        inhibit_on_failure
        TCP_CHECK {
            retry 1
            connect_timeout 1
            connect_port 80
        }
    }
}
```
where the quarum script 
```sh
echo [$(/usr/bin/date +%F.%T)] $1 >> /tmp/quorum.log
[ "$1" = "UP" ] && /usr/sbin/ip addr add 1.1.1.1/32 dev lo ||  /usr/sbin/ip addr del 1.1.1.1/32 dev lo 
```

Reply:
The problem does exist in my tests as shown below. Maybe it's caused by the preemptive  process scheduling for running the quorum script.
![image](https://github.com/iqiyi/dpvs/assets/4437867/e64d3da4-11c1-4d4d-aa80-37cf2d125368)

Reply:
Can you see your keepalived logs when the sequence reversed?
Under normal circumstances, the log "Gained quorum 1+0=1 <= 1 for VS [x.x.x.x]:tcp:80" means quorum_down script excuted, and then new_vs log "Activating healthchecker for service [x.x.x.x]:tcp:80 for VS [x.x.x.x]:tcp:80" means quorum_up script excuted.
Are the logs sequence right?

Reply:
![image](https://github.com/iqiyi/dpvs/assets/4437867/e14c071a-4d64-4148-b323-9dda8b73a7bf)

Reply:
Then sequence in log seems right.
Maybe it's caused by the preemptive process scheduling.

Reply:
The patch solves the problem to some extent but not totally. I'll merge it later, and mark a new problem for the execution sequence problem of the quorum scripts.